llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp

```diff
@@ -33,10 +33,46 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
   const LLT s32 = LLT::scalar(32);
   const LLT s64 = LLT::scalar(64);
 
+  const LLT nxv1s8 = LLT::scalable_vector(1, s8);
+  const LLT nxv2s8 = LLT::scalable_vector(2, s8);
+  const LLT nxv4s8 = LLT::scalable_vector(4, s8);
+  const LLT nxv8s8 = LLT::scalable_vector(8, s8);
+  const LLT nxv16s8 = LLT::scalable_vector(16, s8);
+  const LLT nxv32s8 = LLT::scalable_vector(32, s8);
+  const LLT nxv64s8 = LLT::scalable_vector(64, s8);
+
+  const LLT nxv1s16 = LLT::scalable_vector(1, s16);
+  const LLT nxv2s16 = LLT::scalable_vector(2, s16);
+  const LLT nxv4s16 = LLT::scalable_vector(4, s16);
+  const LLT nxv8s16 = LLT::scalable_vector(8, s16);
+  const LLT nxv16s16 = LLT::scalable_vector(16, s16);
+  const LLT nxv32s16 = LLT::scalable_vector(32, s16);
+
+  const LLT nxv1s32 = LLT::scalable_vector(1, s32);
+  const LLT nxv2s32 = LLT::scalable_vector(2, s32);
+  const LLT nxv4s32 = LLT::scalable_vector(4, s32);
+  const LLT nxv8s32 = LLT::scalable_vector(8, s32);
+  const LLT nxv16s32 = LLT::scalable_vector(16, s32);
+
+  const LLT nxv1s64 = LLT::scalable_vector(1, s64);
+  const LLT nxv2s64 = LLT::scalable_vector(2, s64);
+  const LLT nxv4s64 = LLT::scalable_vector(4, s64);
+  const LLT nxv8s64 = LLT::scalable_vector(8, s64);
+
   using namespace TargetOpcode;
 
-  getActionDefinitionsBuilder({G_ADD, G_SUB, G_AND, G_OR, G_XOR})
-      .legalFor({s32, sXLen})
+  const std::initializer_list<LLT> ALOpLegalScalarTypes = {s32, sXLen};
+  const std::initializer_list<LLT> ALOpLegalVExtendedTypes = {
+      s32,     sXLen,    nxv1s8,   nxv2s8,  nxv4s8,  nxv8s8,
+      nxv16s8, nxv32s8,  nxv64s8,  nxv1s16, nxv2s16, nxv4s16,
+      nxv8s16, nxv16s16, nxv32s16, nxv1s32, nxv2s32, nxv4s32,
+      nxv8s32, nxv16s32, nxv1s64,  nxv2s64, nxv4s64, nxv8s64};
+
+
+  auto ALOpLegalTypes =  ST.hasVInstructions() ? ALOpLegalVExtendedTypes : ALOpLegalScalarTypes; 
+  
+	getActionDefinitionsBuilder({G_ADD, G_SUB, G_AND, G_OR, G_XOR})
+      .legalFor(ALOpLegalTypes) 
```

Comment:
From an earlier review:

> Can we keep the scalar .legalFor separate from the vector one?

Reply:
Right I was trying to keep it by having `.legalFor` scalar and then `.legalIf` vectors, but this wouldn't even work for the `.legalIf(return true)` case

Reply:
Can we use two `legalFor`s?

```
auto AllVecTys = std::initializer_list<LLT>{
    nxv1s8,  nxv2s8,   nxv4s8,  nxv8s8,   nxv16s8,  nxv32s8, nxv64s8, nxv1s16,
    nxv2s16, nxv4s16,  nxv8s16, nxv16s16, nxv32s16, nxv1s32, nxv2s32, nxv4s32,
    nxv8s32, nxv16s32, nxv1s64, nxv2s64,  nxv4s64,  nxv8s64};

 getActionDefinitionsBuilder({G_ADD, G_SUB, G_AND, G_OR, G_XOR})
     .legalFor({s32, sXLen})
     .legalFor(ST.hasVInstructions() ? AllVecTys
                                                           : std::initializer_list<LLT>())
```

In any case, I'd like to have `AllVecTys` exist so it can be reused by future patches. I am confident that we'd like to use those types to legalize other opcodes.

Reply:
> I am confident that we'd like to use those types to legalize other opcodes.

sure then I'll make a `AllVecTys` variable and keep it!